### PR TITLE
Split GBasis into 4 objects. *Breaks API*

### PR DESCRIPTION
**Pull request by matt-chan**
The general idea is described in #28. 

This makes some sense even if we replace the internal implementation of GBasis because it will mean the person changing it in the future is not going to have to replace everything at once. It will be entirely possible to just replace the 1e integrals, or the 2e integrals, or the gb-grid code alone.

Please take a look at the code and see if the following structure (and naming!) is acceptable. The `compute_*` functions will be unchanged from the current implementation, unless someone proposes a new naming for them. 

The structure is similar to QCGrids if you're familiar with that already. The only big change is the unit tests are implemented in Python. It's too much work to move them all over to gtests.

If there's no comments, or if the discussion dies down by Nov 13th (11:59pm EST?) I'll start making the changes! If you need more time to make comments, just let me know! We can extend to Nov 20th.

### CODE REVIEW

**Comment by matt-chan** [comment]
Feel free to add anyone else who you think should review the API.

**Comment by tczorro** [comment]
I am not an expert for gbasis and meanfield structure. I cam help later on with the porcelain parts though.
Chemtools is based on py2-horton. I think it would be a good idea to keep an old gbasis with old api for py2/py3 compatible. We can migrate things to new gbasis if it outperform the old one. I think it is a good practise to put future warning/deprecated warning for the parts we are going to change. So when people develop on the those code, they are informed that these code are changing.

**Comment by matt-chan** [comment]
@tczorro Yeah, I've been keeping  doc/changelog.rst. It's not comprehensive though. There's small changes that get lost. 

As for keeping py2-horton/gbasis... The old code will always exist in git, so people are free to use that if they'd like? Plus the old packages are still published in Conda as well. The only thing is that there's no more maintenance for them. If something breaks... well that's unfortunate but maybe a sign to upgrade to something newer?

**Comment by tczorro** [comment]
Just put a warning like numpy used(Future warning etc) at the place where the api will be removed/changed. So people are prepared. It is annoying and discouraging when the dependencies break your code without notification.

**Comment by matt-chan** [comment]
Hmm yeah that's not too hard to do. Looks like Python has it built into the ``warnings`` module. 

It won't help if you're using something too old though. If we decide to make a change to GBasis and the current master version is "0.1.5", if you're still on "0.1.4", and you upgrade to "0.2.0", then you won't get any warning.

**Comment by tczorro** [comment]
> Hmm yeah that's not too hard to do. Looks like Python has it built into the `warnings` module.
> 
> It won't help if you're using something too old though. If we decide to make a change to GBasis and the current master version is "0.1.5", if you're still on "0.1.4", and you upgrade to "0.2.0", then you won't get any warning.

That's why I mentioned it's better to have it separate. If you are on 0.14, you can only upgrade to 0.15. Then you would see the warning and decide whether to switch to 0.2 or not. When you manually switch to 0.2, you are prepared to handle the compatibility problems. But not upgrade to 0.2 directly and everything's broken. Since I don't know how many api would be broken, I think it is good to put some notes in Readme.

**Comment by matt-chan** [comment]
> That's why I mentioned it's better to have it separate. If you are on 0.14, you can only upgrade to 0.15. Then you would see the warning and decide whether to switch to 0.2 or not.

We can't do that without keeping separate repos or playing some very nasty tricks in conda-build to make sure you don't install 0.1.5 and 0.2.0 at the same time. We're already up to 10+ repos so I'd rather not make a new one every time we make an API breaking change. 

Is it not enough to specify your versions in your own meta.yaml? So when you're making your project, you can specify the library versions like gbasis <=0.1.5 if you know you aren't compatible with anything newer. The semantic versioning (after 1.0?) is supposed to already tell you that going up a major version is API incompatible. I'll do that eventually, but we're changing API so frequently right now that I don't think it's a good idea, unless we want GBasis 9.0 or something by next week...

**Comment by tczorro** [comment]
Well, I think forked it as a new developer version would be fine, but up to you to decide.
Al least for me, I usually don't specify my dependencies' version. So the latest code would be installed automatically.
I think at least, let's put a change log in readme so everyone can easily tract which api is changed at which version.

**Comment by matt-chan** [comment]
Yep, the changelog is already there. It was ``doc/changes.rst``, not ``doc/changelog.rst`` sorry.

It's probably good practice to specify the version number if you're worried about that. Most of the time though, it'll just fail the conda-build/unit tests and you'll know you have to blacklist that version. Less work on your part I think.

Probably better not to use the development fork. It's pretty painful to synchronize the branches, and when someone external makes a bug report/PR, you have to somehow merge it into your newly developed code...

**Comment by PaulWAyers** [comment]
We should see what @FarnazH and @tovrstra say about that. I know that having multiple forks (or something like this) was already something we were planning to do in light have the need for private (developer) and public (released) repos.

**Comment by matt-chan** [comment]
I'm okay with using external forks in theory, but I'd like to see a good workflow with one. I still have flashbacks to the old private/public HORTON repos. Also, it effectively doubles the number of repos to 20+.

Also I'm not sure this actually helps with API breaking changes. It saves us once, (0.1.5 to 0.2.0) but if we make another api breaking change (0.3.0), we'll have to fork again...

**Comment by tovrstra** [comment]
I agree with Matt that other packages using gbasis, which is undergoing API changes, need to specify versions in their dependency list.

Maintaining backward compatibility in the refactoring process is too labour intensive, which cannot be justified by others being just unwilling to specify the version of a dependency. The latter requires no labour at all. If another package wants to use a more recent version of gbasis, this should be a conscious decision and there will then also be a little bit of work to adopt the new API.

This will obviously change when we reach more mature code. Then it starts to make sense to offer API stability. At this stage, I would not bother at all.

**Comment by tovrstra** [comment]
The proposed directory layout looks good. I understand that moving over the massive number of Python tests is too much work at this stage. It needs to be done separately to make this refactoring feasible.

Many unit tests are currently not written in a maintainable way. The test code contains many hardcoded reference results (Numpy arrays) with a massive amount of repetitive code. Literally translating that to C++ is too painful. Ideally, such reference results should be stored into simple data files, which can be easily processed by just a limited amount of testing code. Once these data files are made, the large number of tests can be transferred to gtest (or refactored in other ways in future) with less effort. Again, this is better done in another pull request.

**Comment by tovrstra** [comment]
Regarding the multiple repos, we should just make sure there is one and only clean central (and thus public) repository where we impose some order. Anyone is then still free to fork personal repositories, which they can manage themselves, e.g. to prepare new pull requests or to store their unpublished work. In the public repo there should only be 

I would avoid a second central private repository that is also actively managed and used by many people. Keeping two central repos in sync somehow, with a delay on unpublished work, is a nightmare. The main drawback from such an approach is that new features in the private repo are not split into manageable pull requests. Instead, the private repo would contain a master branch with many new features in various degrees of completion and quality, which are not easily taken out and merged one by one into a public repo. As Matt said, this is what went wrong in the development of HORTON 2.

**Comment by matt-chan** [comment]
I'm making a few other changes to the API. I think it makes more sense to call the files 2/4 instead of 1/2, since all the classes inside are named as such.

Also, I've been working with the CMake scripts from GQCP (Patrick's code) and I like their organization better. There's lots of project-specific stuff in the way we do it on qcgrids, whereas everything project specific is in a environment file for GQCP.

**Comment by matt-chan** [comment]
It appears Cython isn't able to store pxds separately from the pyx files... sigh. The attribute types aren't don't get resolved if you keep them separately. We're going to have one big directory containing everything Cython/Python I think.

**Comment by matt-chan** [comment]
Actually, it might be possible. It appears the Cython build tools have changed significantly since we last updated the setup.py scripts. Our method using `cmd_class` is actually no longer in the Cython docs. Everything is done using `cythonize` now. In addition, the `Extension` instances are acceptable arguments to it, or even factories for `Extension` (probably overkill for our purposes). 

I'll see what I can dig up...

**Comment by matt-chan** [comment]
Yep, I've got the pxd search working in the latest commit. What a pita. 

A few other things to go. I still need to polish the python builds, update the tests, and then add conda to the python package. Then I have to get Travis working with the new structure. Finally, I need to write some docs about the new structure and about how to change it according to the wishlish issues. 

I was considering splitting the C++ shared lib into 4 objects as well, but I don't think it's worth it. When we get to the point where we want to add wishlist issues (e.g. remove the 1e ints from C++), then we can simply just delete the ones source folder and the functions from gbasis.cpp.

**Comment by matt-chan** [comment]
Okay that commit is the last of the code changes. Everything else is conda, docs etc. Could whoever is interested review the PR please? If there's no comments by Tues Nov 27th 11:59pm (or a request to extend the window), I'll merge it in. 

In particular, I'm not sure if it still makes sense to call things cext1/2 etc. when internally it's referred to as ints2/4. Please let me know if I should change it from how it is right now.

**Comment by matt-chan** [comment]
I've changed the conda package names to cxx-gbasis and python-gbasis. I figured since the new version is pretty incompatible with the older one (cxx-gbasis, which would have been called gbasis without the rename) and doesn't even have a python interface, it would be easier to move it to a new conda version.

**Comment by matt-chan** [comment]
I've split off the C++ lib version number from the Python module. It was too hard to get conda building with both of them on non-release builds without it. Even though the packages aren't published officially, a proper version number is needed for the local channel in order to install it, and the tag system doesn't work well if that particular commit isn't tagged (it falls back to the last tag I think). 

Long story short, if you make changes to the C++ library in the future, you've got to crank the version number in ``cpp/conda.recipe/meta.yaml`` manually.

**Comment by FarnazH** [comment]
@matt-chan are you done with your split?

**Comment by PaulWAyers** [comment]
We should think about this. The main thing is that there are shared features (mainly Boys function) between 1-electron electron-nuclear integrals and 2-electron integrals.  The split is good but there may be some bits that are either copied or shared functions.

**Comment by matt-chan** [comment]
@FarnazH Sort of. The code is done and passes unit tests, but I haven't finished documentation and some of the linters are still failing. I left it off for a bit to prioritize some stuff with another project. I'd actually like some code review on this if you guys have time down there =).

Also, if you guys want to change the structure of the code, please let me know asap! It would mean I don't have to fix the linters and write documentation twice.

**Comment by tovrstra** [comment]
Merging is fine by me. We are indeed thinking of some restructuring, for which we'll make issues. Initially, the main focus of this C++ implementation would be the evaluation of properties on integration grids. Other features overlap a lot with existing libraries (libint, libcint, libmint, libreta, ...), so it is not critical to keep the redundant features alive while restructuring gbasis.
